File: 자료구조/HashTable/HashTableWithInnerArray.java
public class HashTableWithInnerArray {
    Slot[] dataArray;
    int  size;
    public HashTableWithInnerArray(int size){
       this.size = size;
       dataArray  =  new Slot[size];
    }

    private class Slot{
        String key;
        String value;
        public Slot(String key, String value) {
            this.key = key;
            this.value = value;
        }
    }

    private int hashFun(String inputKey){
        int idx=0;
        for(int i=0 ; i<inputKey.length() ; i++){
            idx += (int)(inputKey.charAt(i));
        }
        return idx % size;
    }

    public boolean inputData(String inputKey, String inputValeu){
        int idx = hashFun(inputKey);
        int cnt = 0;

        while(this.dataArray[idx] != null){
            cnt ++;
            if (cnt > size){
                return false;
            }
            if(this.dataArray[idx].key.equals(inputKey)){
                this.dataArray[idx].value = inputValeu;
                return true;
            }
            if(idx == size-1){
                idx = 0;
            }else{
                idx += 1;
            }
        }
        this.dataArray[idx] = new Slot(inputKey, inputValeu);
        return true;
    }

    public String getData(String inputKey){
        int idx = hashFun(inputKey);
        int cnt = 0;

        while(this.dataArray[idx]!=null){
            cnt ++;
            if (cnt > size){
                return null;
            }
            if(this.dataArray[idx].key.equals( inputKey)){
                return this.dataArray[idx].value;
            }
            if(idx == size-1){
                idx = 0;
            }else{
                idx += 1;
            }
        }
        return null;
    }

    public static void main(String args[]){
        HashTableWithInnerArray hash = new HashTableWithInnerArray(3);
        System.out.println(hash.inputData("박상우","야호"));//true
        System.out.println(hash.inputData("박상우1","야호1"));//true
        System.out.println(hash.inputData("박상우2","야호2"));//true
        System.out.println(hash.inputData("박상우2","야호22"));//true
        System.out.println(hash.inputData("박상우3","야호3"));//false
        System.out.println("=======");
        System.out.println(hash.getData("박상우")); //야호
        System.out.println(hash.getData("박상우1"));//야호
        System.out.println(hash.getData("박상우3"));//null
        System.out.println(hash.getData("박상우2"));//야호
        System.out.println(hash.getData("박상우4"));//null

    }
}

File: 자료구조/HashTable/HashTableWithInnerArray.kt
fun main(){
    val hash = HashTableWithInnerArray(5).apply {
        println(inputData("박상우1","야호1"))
        println(inputData("박상우2","야호2"))
        println(inputData("박상우3","야호3"))
        println(inputData("박상우4","야호4"))
        println(inputData("박상우5","야호5"))
        println(inputData("박상우6","야호3"))
        println(inputData("박상우5","데이터 바뀐것."))

        println()

        println(searchData("박상우1")) //야호1
        println(searchData("박상우2")) //야호1
        println(searchData("박상우3")) //야호1
        println(searchData("박상우4")) //야호1
        println(searchData("박상우6")) //null
        println(searchData("박상우5")) //데이터 바뀐것.
    }
}

class HashTableWithInnerArray(val size : Int){
    var dataList : Array<Slot?>
    init {
        this.dataList = Array<Slot?>(size,{null})
    }

    class Slot(val key : String, var value : String){ }


    fun hashTable(inputKey: String):Int {
        var idx =0
        for (i in inputKey.indices) {
            idx += inputKey[i].toInt()
        }
        return idx % this.size
    }

    fun inputData(inputKey: String, inputValue : String) : Boolean{
        var idx = hashTable(inputKey)
        var cnt = 0
        while(this.dataList[idx] != null){
            if (++cnt > this.size){
                return false
            }
            if(this.dataList[idx]!!.key == inputKey){
                this.dataList[idx]!!.value = inputValue
                return true
            }
            idx ++
            if(idx == this.size)
                idx = 0
        }
        this.dataList[idx] = Slot(inputKey,inputValue)
        return true
    }

    fun searchData(inputKey: String) : String? {
        var idx = hashTable(inputKey)
        var cnt = 0
        while(this.dataList[idx] != null){
            cnt++
            if (cnt>size) {
                return null
            }
            if(this.dataList[idx]!!.key == inputKey){
                return this.dataList[idx]!!.value
            }
        }
        return null
    }
}